### ISSUE TITLE
mbedtls_x509write_crt_set_subject_name, mbedtls_x509_dn_gets fail to handle UTF8 multibyte properly

### ISSUE THREAD

**Issue by mman** [ARMmbed/mbedtls#3413]
I'm on mbedtls-2.16.6.

I generate a self signed cert and I pass a utf-8 encoded multibyte string into certificate subject and issuer using:

```
mbedtls_x509write_crt_set_subject_name 
```

I then retrieve it back using:

```
mbedtls_x509_dn_gets
```

It works great for ascii, however when I pass in a string containing the sequence below representing an apostrophe in UTF8 (')

```
0xe2 0x80 0x99
```

It will not get parsed out correctly. What I actually get back is where the last byte of multibyte sequence is replaced with question mark (?)

```
0xe2 0x80 0x3f
```

I do not understand enough of ASN.1 to troubleshoot this further, however I was able to figure out that ``mbedtls_x509_dn_gets`` is actually trying to parse out tag 12 that should be utf8.

**Comment by mman** [comment]
I have tried to apply the patch https://github.com/ARMmbed/mbedtls/pull/3326/files against 2.6.16, but that patch fails to decode utf-8 multibyte stored in the subject field.

Instead it simply replaces all occurrences of utf-8 control chars with question mark therefore masking the utf-8 completely. Not a good fix IMHO.

Effectively turning a string ``Hello'``

```
Hello\xe2\x80\x99
```

into 

```
Hello???
```

**Comment by danh-arm** [comment]
Hi @mman. It's better to ask questions like this on the Mbed TLS developer list: 
https://lists.trustedfirmware.org/mailman/listinfo/mbed-tls

This issue tracker is now only for (confirmed) bugs and enhancements. It sounds like what you're asking for is to make this code UTF-8 aware but even then there will be limitations to the characters that can be printed out. If there's consensus on the list that this would have value then an enhancement task can be created.

**Comment by mman** [comment]
@danh-arm Moved the discussion over, thanks for looking into it.

**Comment by gilles-peskine-arm** [comment]
This is the same bug as https://github.com/ARMmbed/mbedtls/issues/3303, isn't it? [This bug is fixed in the `development` branch](https://github.com/ARMmbed/mbedtls/pull/3326) and the fix will be in the next release (2.23).

In https://github.com/ARMmbed/mbedtls/pull/3326#issuecomment-633544478 I wrote that this didn't need to be backported because it's “a minor bug in a function that is for debugging and logging purposes”, but I was thinking of `mbedtls_x509_crt_info`, not `mbedtls_x509_dn_gets`. Should we backport the fix after all?

**Comment by mman** [comment]
@gilles-peskine-arm The problem is that the fix for #3303 is kind of broken IMHO. Instead of correctly passing out three bytes of UTF-8 sequence it actually replaces all control chars with ? effectively turning a string like this ``SuperHero's Device`` into ``SuperHero???s Device`` because the apostrophe in the original name is represented as UTF-8 multibyte sequence ``\xe2\x80\x99``.

All this is unnecessary I think because the underlying ASN representation is tagged as UTF-8.

I fail to see the logic behind replacing the control chars completely.

**Comment by gilles-peskine-arm** [comment]
@mman Ah, indeed. I'll reply to [your message on the mailing list](https://lists.trustedfirmware.org/pipermail/mbed-tls/2020-June/000093.html).
